File: epi_judge_java/epi/IsTreeSymmetric.java
package epi;
import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;
public class IsTreeSymmetric {
  @EpiTest(testDataFile = "is_tree_symmetric.tsv")

  public static boolean isSymmetric(BinaryTreeNode<Integer> tree) {
	  // Implement this placeholder.
	  // algo: for each of the subtrees (subtree1, subtree2) check if subtree2.left = subtree2.right
	  if (tree == null) {
		  return true;
	  }
	  return isSymmetric(tree.left, tree.right);
  }

	static boolean isSymmetric(BinaryTreeNode<Integer> leftTree, BinaryTreeNode<Integer> rightTree) {
		// if one is null the other is not return false;
		if (leftTree == null && rightTree == null) {
			return true;
		}
		if (leftTree != null && rightTree != null) {
			if (!leftTree.data.equals(rightTree.data)) {
				return false;
			}
			return 	isSymmetric(leftTree.right, rightTree.left) &&
					isSymmetric(leftTree.left, rightTree.right);
		}
		return false;
	}

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "IsTreeSymmetric.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}

File: epi_judge_python/is_tree_symmetric.py
from binary_tree_node import BinaryTreeNode
from test_framework import generic_test


# 09:15
# 09:22 finished writing code
# 09:23 finished debugging. all working 1st shot :)
def is_symmetric(tree: BinaryTreeNode) -> bool:
    if not tree:
        return True
    return check_symmetric(tree.left, tree.right)


def check_symmetric(left: BinaryTreeNode, right: BinaryTreeNode) -> bool:
    if not left and not right:
        return True
    elif left and right:
        # compare opposite child nodes
        return left.data == right.data and \
               check_symmetric(left.left, right.right) and \
               check_symmetric(left.right, right.left)
    else:
        # only one of the child nodes is missing
        return False


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('is_tree_symmetric.py',
                                       'is_tree_symmetric.tsv', is_symmetric))
